specs/derivation.md

```diff
@@ -500,7 +501,51 @@ On L1, the blob data is represented as a polynomial of points, each in range `[0
 just below the capacity of a `uint256`: approximately `254.857` bits of data.
 Before proceeding with processing, the data is transformed to turn it into a continuous byte string.
 
-TODO: blob encoding format. Version + length-prefix + maximize utilization of the data.
+##### Blob encoding
+
+Each blob in a [EIP-4844] transaction really consists of `FIELD_ELEMENTS_PER_BLOB = 4096` field elements.
+
+Each field element is a number in a prime field of
+`BLS_MODULUS = 52435875175126190479447740508185965837690552500527637822603658699938581184513`.
+This number does not represent a full `uint256`: `math.log2(BLS_MODULUS) = 254.8570894...`
+
+The [L1 consensus-specs](https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/polynomial-commitments.md)
+describe the encoding of this polynomial.
+The field elements are encoded as big-endian integers (`KZG_ENDIANNESS = big`).
+
+To save computational overhead, only `254` bits per field element are used for rollup data.
+
+Data is encoded, 4 field elements, `127` bytes of rollup data, at a time:
+
+```text
+# Allocate 4 field elements (big-endian).
+# Read 31 bytes into the low 31 bytes of each.
+<----- element 0 -----><----- element 1 -----><----- element 2 -----><----- element 3 ----->
+| byte A |  read(31)  || byte B |  read(31)  || byte C |  read(31)  || byte D |  read(31)  |
```

Comment:
Would we be able to design a format which could stream in sooner rather than having to go over 127 bytes? I guess it's not a huge buffer, but it's nice to immediately write stuff instead of having to set bits after the fact

Reply:
Good point, we can change it to read the 3 bytes from rollup data first, so only those bytes have to be buffered.

Reply:
Tried the above, but then proving of the rollup-data prefix bytes (version and length) becomes harder, due to those first 3 bytes being split between 4 field elements.

Instead, we can also do: https://github.com/ethereum-optimism/optimism/pull/8745
This reduces it to only buffer 32 bytes, while allowing proving of the start with just the first field-element, and writing of 1 field element at a time.

And expresses it with pseudocode, and clarifies what's rollup-data and blob-data, as requested by you and @tynes